hax/hax/motr/planner.py

```diff
@@ -0,0 +1,297 @@
+        # additional chunks of memory. That's why group id should be wrapped
+        # back to zero manually.
+        if new_value > MAX_GROUP_ID:
+            new_value = 0
+        return new_value
+
+    def _inc_group(self):
+        ''' Increases the currently active group. The method is invoked by
+            WorkPlanner when all the commands from the current group have
+            already been processed.
+            Assumes that b_lock is acquired already.
+        '''
+        state = self.state
+        cur_group_id = state.current_group_id
+        change_next_group = state.next_group_id == cur_group_id
+
+        state.current_group_id = self._get_increased_group(cur_group_id)
+
+        if change_next_group:
+            state.next_group_id = state.current_group_id
+            state.next_group_commands = set()
+
+    def notify_finished(self, command: BaseMessage) -> None:
+        ''' The method must be invoked by the worker thread when the command
+            is executed.
+        '''
+
+        with self.b_lock:
+            state = self.state
+            state.active_commands.remove(command)
+            LOG.log(TRACE, '[WP]Cmd %s removed. Current state: %s', command,
+                    state)
+
+            if state.active_commands:
+                return
+            for c in self.backlog:
+                if c.group == state.current_group_id:
+                    return
+            for c in state.taken_commands:
+                if c.group == state.current_group_id:
+                    return
+            # if we're here, command was the only one belonging to group
+            self._inc_group()
+            LOG.log(TRACE, '[WP]Active group changed to %s',
+                    state.current_group_id)
+            # The group changed, let's unblock those who are waiting for
+            # this group
+            self.b_lock.notifyAll()
+
+    def _should_increase_group(self, cmd: BaseMessage) -> bool:
+        ''' Predicate function. Returns True if and only if cmd command CANNOT
+            be added to group number next_group_id.
+            Assumes that b_lock is acquired already.
+        '''
+        def has(cmd_type: Type[BaseMessage]) -> bool:
+            ''' Checks if the group being currently formed (i.e. next_group)
+                contains a message of the given type.
+            '''
+            return cmd_type in self.state.next_group_commands
+
+        if not self.state.next_group_commands:
+            # current group is empty -> join it freely
+            return False
+        if isinstance(cmd, HaNvecGetEvent):
+            # HaNvecGetEvent can be done in parallel to any other commands.
+            # No need to form the new group for it.
+            return False
+        if isinstance(cmd, AnyEntrypointRequest):
+            # if the current group has a BroadcastHAStates request,
+            # then this entrypoint request should be placed to a next group.
```

Comment:
So broadcast and entrypoint requests will not be part of the same group. Which seems to be okay but is it possible that one of the consumer thread picked up the last cmd in the current thread and other thread starts executing the next_group as it does not find any more commands in the current group?

Reply:
The thread which picked up a command from `next_group_id` will be allowed to process it if and only if `current_group_id` becomes equal to the command's group id.
Current group id gets changed when all these criteria are true:
1. in backlog there is no command with `group_id == current_group_id` (backlog is used as a source for `get_next_command()`)
2. in `taken_commands` list there is no command `group_id == current_group_id` (which means that there is no ConsumerThread that has already taken a command from the backlog with such an id but has not started processing it yet).
3. in `active_commands` list there is no command `group_id == current_group_id` (this list contains the commands being processed at this moment, so WorkPlanner has already allowed their execution)